[quantization] Store bias in PackedLinearWeight struct in fbgemm

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25626 [quantization] Store bias in PackedConvWeight in fbgemm
* **#25428 [quantization] Store bias in PackedLinearWeight struct in fbgemm**
* #25678 [quantization] Rename FBGEMM quantized operators to generic quantized ops

Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

supriyar committed:
[quantization] Store bias in PackedLinearWeight struct in fbgemm

Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

supriyar commented:
@jamesr66a Is there a reason why the entire function signature is specified for the linear op but not for the conv op? `quantized::fbgemm_linear(Tensor X, Tensor W_prepack, float Y_scale_i, int Y_zero_point_i) -> Tensor Y"` vs `quantized::fbgemm_conv2d`

I'm seeing jenkins failure because the function signature for linear changed from 5 to 4 arguments.
https://circleci.com/gh/pytorch/pytorch/2632323?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

dzhulgakov commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -119,7 +119,7 @@ def __init__(self, in_features, out_features, bias_=True):
             [out_features, in_features], scale=1, zero_point=0, dtype=torch.qint8)
 
-        self.set_weight(qweight)
+        self.set_weight_bias(qweight, self.bias)
```
so the problem is that you already quantized the bias above, right?

jianyuh replied:
Right: maybe use c10::optional<Tensor> bias to make bias optional.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -93,18 +92,25 @@ class QLinearInt8 final : public torch::OperatorKernel {
+      // Temporary: Quantize bias
+      at::Tensor qbias = bias;
+      if (!bias.is_quantized()) {
+        qbias = at::quantize_linear(
```
this is different logic from what was before this diff. Note that bias is already quantized, but with wrong scale/zp. Therefore, here if it's quantized we'd need to requantize it with correct scale.

Btw, if our unittests didn't catch it - we have crappy unittests.

supriyar replied:
Dequantizing and re-quantizing the bias is causing various unit tests to fail, with either mismatching output or ASAN errors. I pinged @jianyuh to help with this. Also cc @dskhudia .

jianyuh replied:
For bias in `fbgemm_linear`, it is already quantized outside and the `bias` here should be `torch.qint32` data type. Check https://github.com/pytorch/pytorch/blob/master/test/test_quantized.py#L812-L813. So we don't need to apply `quantize_linear` inside this operator

supriyar replied:
This is the same as the previous code in https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/linear.py#L133, which always requantizes the bias. In the commented out code below, I'm trying to do that but it causes unit tests to fail which is very weird.

dskhudia replied:
This part of the code needs to handle per channel quantization. See use of qscheme above in the code. This should make your ASAN errors go away. 
`qbias = at::quantize_linear(at::dequantize(bias), pack_ptr.w_scale[0] * input_scale_float, 0, kQInt32);`

I still see some output mismatch even if this issue is not there. I am not yet sure why.

supriyar replied:
@dskhudia good catch. I'll take that into account. However this fails even without per channel quantization enabled.

dskhudia replied:
Added code needed that check. Once we handle per channel case,  the check would be unnecessary.

dskhudia replied:
Please see my comment below for the other data mismatch issue.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp:
```diff
@@ -94,7 +103,7 @@ class QLinearPackWeightInt8 final : public c10::OperatorKernel {
 
 static auto registry = c10::RegisterOperators().op(
-    "quantized::linear_prepack(Tensor W) -> Tensor W_prepack",
+    "quantized::linear_prepack(Tensor W, Tensor? B) -> Tensor W_prepack",
```
can you specify default value? that way we don't need to pass it in python all the time as None

dzhulgakov commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -167,7 +161,7 @@ def __getstate__(self):
+        self.set_weight_bias(state_dict[prefix + 'weight'], state_dict[prefix + 'bias'])
         state_dict.pop(prefix + 'weight')
 
         self.bias = state_dict[prefix + 'bias']
```
you can remove this now, right?

dzhulgakov commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -188,7 +182,7 @@ def __setstate__(self, state):
         self.in_features = state[0]
         self.out_features = state[1]
         self.bias = state[2]
```
can be removed?

supriyar replied:
Yeah, I'll remove the bias

dzhulgakov commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -197,8 +191,9 @@ def __setstate__(self, state):
 
-    def set_weight(self, w):
-        self._packed_weight = torch.ops.quantized.linear_prepack(w)
+    def set_weight_bias(self, w, b):
```
if you still want to keep self.bias for convenience (though we shouldn't really) - let's put this assignment here. But I'm bothered by this discrepancy between bias and weight in unpacking

raghuramank100 replied:
Have a similar concern. Lets eliminate self.bias as a member and only have packed_params.
Instead have two functions: 
```python
def weight(self):
        (w,b) =  torch.ops.quantized.linear_unpack(self._packed_params)
        return w

def bias(self):
        (w,b) =  torch.ops.quantized.linear_unpack(self._packed_params)
        return b
```

dzhulgakov commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -197,8 +191,9 @@ def __setstate__(self, state):
     def weight(self):
         return torch.ops.quantized.linear_unpack(self._packed_weight)
```
maybe make unpack return both bias and weight then?

supriyar replied:
It gets a little messy because bias is an optional tensor in this module and that is causing runtime error  `RuntimeError: sizes() called on undefined Tensor` when trying to call `torch::jit::script::Module::save`
We can perhaps add this in later if required? For now save bias in the module itself when it is set.

supriyar replied:
Resolved this, bias was being set to undefined when there was no bias. I made it an optional Tensor as opposed to at::Tensor and it fixed the error.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -166,10 +172,10 @@ class QLinearInt8 final : public torch::OperatorKernel {
 static auto registry =
     torch::RegisterOperators()
-        .op("quantized::linear(Tensor X, Tensor W_prepack, Tensor? b, float Y_scale_i, int Y_zero_point_i) -> Tensor Y",
+        .op("quantized::linear(Tensor X, Tensor W_prepack, float Y_scale_i, int Y_zero_point_i) -> Tensor Y",
```
@jamesr66a - just to confirm, we don't ship this anywhere so it's fine to change the signature, right?

jamesr66a replied:
Yeah I *believe* it's okay but if we can pull operator stats from somewhere to confirm that would be safest

jianyuh commented on test/test_jit.py:
```diff
@@ -14513,15 +14513,15 @@ def forward(self):
             def __setstate__(self, state):
                 self._packed_weight.set_(
-                    torch.ops.quantized.linear_prepack(state))
+                    torch.ops.quantized.linear_prepack(state, None))
```
If we have `c10::optional<Tensor> bias`, we don't need to add None here and provides the backward compatibility.

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

jamesr66a reviewed (changes requested):
ROCM failure is true-positive. The function signature at qlinear.cpp:218 needs to be updated

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

dskhudia commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -107,15 +106,27 @@ class QLinearInt8 final : public torch::OperatorKernel {
-      // TODO: contiguous is called for further jit optimizations.
-      auto bias_contig = bias_vec.contiguous();
+      auto bias_contig = qbias.contiguous();
       bias_ptr =
```
@supriyar @dzhulgakov 
We are grabbing a data pointer of a tensor in local scope here. Fixing this should fix the issue. I tried it locally and it worked.

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

jgong5 commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -108,14 +107,34 @@ class QLinearInt8 final : public torch::OperatorKernel {
+    at::Tensor qbias;
+    if (pack_ptr.bias.has_value()) {
+      at::Tensor bias = pack_ptr.bias.value();
+      // Temporary: Quantize bias
```
Will you consider to avoid requantizing bias every time for better performance? Caching and compare the scale/zp of the input tensor should work fine.

supriyar replied:
I think this is just a temporary solution and it will be replaced. The plan is to quantize bias dynamically as a post operation step after linear/conv. cc @dskhudia

raghuramank100 replied:
Bias will need to be requantized every time in eager mode, this is because we do not know the input tensor's scale and zero-point ahead of time. When we have a static graph, we can optimize this further.

dskhudia commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -108,14 +107,34 @@ class QLinearInt8 final : public torch::OperatorKernel {
+        }
+        qbias = quantize_linear_per_channel_cpu(
+            at::dequantize(bias), bias_scale, bias_zp, axis, kQInt32);
+      }
```
Add an else part here with a torch check that fails. You may want to move qbias = bias (line 114) in the else here.

dskhudia commented:
I have approved it. Please take care of the tests.

dskhudia commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -108,14 +107,34 @@ class QLinearInt8 final : public torch::OperatorKernel {
+          bias_scale.data_ptr<double>()[i] =
+              pack_ptr.w_scale[i] * input_scale_float;
+        }
+        qbias = quantize_linear_per_channel_cpu(
```
Use quantize_linear_per_channel instead of quantize_linear_per_channel_cpu. quantize_linear_per_channel is the api exposed in Python frontend.

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp:
```diff
@@ -97,6 +106,7 @@ class QLinearPackWeightInt8 final : public c10::OperatorKernel {
             /*ld=*/K,
             /*pmat=*/nullptr, // PackBMatrix manages ownership of pmat
             /*groups=*/1),
+        bias_contig,
```
How does this work when bias is None?

supriyar replied:
It is stored as an optional tensor - similar to the current linear op that has bias as an input argument. So if it is none then that gets taken care of.

raghuramank100 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -146,18 +140,20 @@ def forward(self, x):
     def _save_to_state_dict(self, destination, prefix, keep_vars):
         super(Linear, self)._save_to_state_dict(destination, prefix, keep_vars)
-        destination[prefix + 'weight'] = self.weight()
+        (w, b) = self.weight()
```
Can we have a seperate function self.bias() to obtain the bias instead? This way the function names map to individual attributes.

supriyar replied:
@dzhulgakov suggested to update existing `weight()` function to return bias as well. Maybe I can rename it to weight_bias() instead?

raghuramank100 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -167,10 +163,8 @@ def __getstate__(self):
     def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                               missing_keys, unexpected_keys, error_msgs):
-        self.set_weight(state_dict[prefix + 'weight'])
+        self.set_weight_bias(state_dict[prefix + 'weight'], state_dict[prefix + 'bias'])
```
Similar comment, lets have individual functions for weight and bias

supriyar replied:
The set_weight_bias function calls the prepack method, which needs both weight and bias as arguments. I don't think it will help to separate them out.

raghuramank100 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -197,8 +190,10 @@ def __setstate__(self, state):
-        self._packed_weight = torch.ops.quantized.linear_prepack(w)
+    def set_weight_bias(self, w, b):
+        # type: (torch.Tensor, Optional[torch.Tensor]) -> None
+        self._packed_weight = torch.ops.quantized.linear_prepack(w, b)
```
Can we call this _packed_weight_and_bias so that the meaning is clear?

supriyar replied:
This is referring more to the `PackedLinearWeight` struct in FBGEMM rather than the actual weight. Also we aren't actually packing the bias with the weight, just adding to the struct. I think changing it to `_packed_weight_and_bias` may cause some misunderstanding.

jamesr66a replied:
IMO not indicating the presence of the bias at all is worse than someone misunderstanding whether bias is packed or not. Maybe named this something like `_opaque_params`?

supriyar replied:
@jamesr66a @raghuramank100 I'm renaming this to `_packed_params` instead.

raghuramank100 commented:
We also need to make similar changes to: https://github.com/pytorch/pytorch/blob/9d06a984f866289c2acb28a379f62378c5e70454/torch/nn/quantized/functional.py

raghuramank100 reviewed (changes requested):
Looks great, a few suggested changes.

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

supriyar commented:
Removed self.bias from the modules. Added `bias()` and `weight()` functions. Also renamed and updated the API for `fbgemm_linear_dynamic`
cc @jianyuh @raghuramank100

supriyar committed:
Update on "[quantization] Store bias in PackedLinearWeight struct in fbgemm"


Added bias as an optional param to the quantized_linear_prepack function.
Bias is quantized during runtime using input scale and weight scale.

Differential Revision: [D17121304](https://our.internmc.facebook.com/intern/diff/D17121304/)

jianyuh commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.h:
```diff
@@ -16,6 +16,7 @@
 // Note that in JIT mode we can think of a way to fuse col_offsets with bias.
 struct FBGEMM_API PackedLinearWeight {
   std::unique_ptr<fbgemm::PackBMatrix<int8_t>> w;
+  c10::optional<at::Tensor> bias;
```
Another option is to just store a std::vector<int32_t> bias instead of at::Tensor for bias. But the drawback is that we are changing the type of bias from int32_t to float, to resolve the scaling factor unknown issue for dynamic batching during the eager mode. cc @dskhudia . I am OK with the current design.

jianyuh commented:
> Removed self.bias from the modules. Added `bias()` and `weight()` functions. Also renamed and updated the API for `fbgemm_linear_dynamic`
> cc @jianyuh @raghuramank100

The updated API for `fbgemm_linear_dynamic` looks good to me!
